[Bug 16686] libfoundation: Better __MCNativeChar_CheckedFold() accuracy

peter-b opened the pull request:
This patch modifies `__MCNativeChar_CheckedFold()` to return a more
accurate result.  It now looks up the input char in a constant bit
array in order to determine whether it needs to be folded.  The
constant bit array is only 32 bytes, which means it fits into half a
cache line.  Because `__MCNativeChar_CheckedFold()` now only calls
back to `__MCNativeChar_Fold()` if necessary, searching strings for
this change should provide a speedup for case-sensitive string
operations on inputs that contain a large amount of digits or
punctuation.

peter-b commented:
Running the benchmarks in release mode for this changeset make for seriously weird results.  This patch _only_ touches `__MCNativeChar_CheckedFold()`, which is provably not **ever** called for "Exact" native string operations -- but the speed of some "Exact" ops is much slower with this change applied.  What's going on? :confounded: 

[perf-native-checked-fold.txt](https://github.com/livecode/livecode/files/68215/perf-native-checked-fold.txt)

runrevmark commented:
The operations that are affected are Exact operations which use ForwardScan to find the needle in the haystack (as opposed to just a single character). Both Offset and Contains dispatch directly to ForwardScan, which does a dynamic choice based on p_options. The only thing I can think of is the slightly larger code size of the new CheckedFold operation has changed the compilers choice of when to inline, or increased the amount of code between the entry to ForwardScan and dispatch to the Unfolded path at the very end. The logic probably should be:

```
if (options is unfolded)
  return unfolded op

... do general path which optimizes folded ops if it can ...
```

runrevmark commented:
Note: When doing the optimization I found how __MCNativeOp_IsFolded was implemented had a huge effect on performance. The comparison was the quickest, bit operations slowed things down significantly.

runrevmark commented:
Actually, I wonder if the code path should look something like this:

```
if (is_unfolded)
{
unfolded:
  return unfolded_op
}
else
{
  if (length < prefold_limit)
  {
    if (!checkfolded(needle))
      goto unfolded

   return prefolded_op
  }
  else
    return folded_op
}
```

The idea here is that 'exact' is the fastest so should be dealt with as likely and first. Then, the next fastest thing is a caseless small needle, so do that next - reusing the unfolded path. The slowest thing is general folding case, so make sure that is least likely and last.

Also, marking certain things as force_inline might help...

livecodefraser commented:
holiday (n): a period of one or more days taken off work by an employee for leisure.

runrevmark commented:
That definition hinges on what one defines as 'leisure' ;)

Sent from my iPhone

> On 21 Dec 2015, at 09:39, Fraser Gordon notifications@github.com wrote:
> 
> holiday (n): a period of one or more days taken off work by an employee for leisure.
> 
> —
> Reply to this email directly or view it on GitHub.

peter-b commented:
@runrevmark The changes we went through yesterday seem to have fixed the performance regression.

[perf-native-checked-fold-2.txt](https://github.com/livecode/livecode/files/79685/perf-native-checked-fold-2.txt)

peter-b committed:
[Bug 16686] Add release note

peter-b committed:
[Perf] Refactor __MCNativeOp_Scan() to improve optimisation

The changes in commit 7f4fd8f made it harder for the compiler to
successfully optimise __MCNativeOp_Scan() in release builds, causing a
substantial slowdown in unfolded scanning.  This patch rearranges
__MCNativeOp_Scan() in order to facilitate optimisation, and restores
the original performance.

peter-b committed:
[Perf] libfoundation: Better __MCNativeChar_CheckedFold() accuracy

This patch modifies `__MCNativeChar_CheckedFold()` to return a more
accurate result.  It now looks up the input char in a constant bit
array in order to determine whether it needs to be folded.  The
constant bit array is only 32 bytes, which means it fits into half a
cache line.  Because `__MCNativeChar_CheckedFold()` now only calls
back to `__MCNativeChar_Fold()` if necessary, searching strings for
this change should provide a speedup for case-sensitive string
operations on inputs that contain a large amount of digits or
punctuation.

peter-b commented:
Updated the patch slightly to fix a stupid error that broke the testsuite.

[perf-native-checked-fold-3.txt](https://github.com/livecode/livecode/files/79794/perf-native-checked-fold-3.txt)

runrevmark commented:
@livecode-vulcan review ok da424e0

livecode-vulcan commented:
💙 review by @runrevmark ok da424e0502d00f5bcfe0ace9f13fb85e5713825b

livecode-vulcan commented:
😎 test success da424e0502d00f5bcfe0ace9f13fb85e5713825b
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1539)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/1558)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1535)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1546)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/1558)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1559)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1561)
